int/sni/sni.go

```diff
@@ -0,0 +1,171 @@
+package sni
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/massalabs/station/pkg/certificate"
+	"github.com/massalabs/station/pkg/logger"
+)
+
+var ErrInvalidArgument = errors.New("invalid argument")
+
+const (
+	privateKeySizeInBits  = 2048
+	loggerPrefix          = "SNI -"
+	caCertificateFileName = "rootCA.pem"
+	caPrivateKeyFileName  = "rootCA-key.pem"
+)
+
+// hashServerName returns a unique identifier for the server, hashing the server's name with the current time stamp.
+// The function returns an error if the server name is empty.
+func hashServerName(serverName string) ([]byte, error) {
+	if len(serverName) == 0 {
+		return nil, fmt.Errorf("%w: server name is empty", ErrInvalidArgument)
+	}
+
+	now, err := time.Now().MarshalBinary()
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal current time: %w", err)
+	}
+
+	return sha256.New().Sum(append([]byte(serverName), now...)), nil
+}
+
+// createCertificateTemplate builds a x509 certificate template using the server name and a unique ID.
+// DNSNames is set to the given server name, the serial number is set to the unique site name id,
+// NotBefore is set to the current time, NotAfter is set to 1 day after the current time,
+// and the organization is set to "station dynamically generated".
+//
+// The function returns an error if the server name or the unique site name id are empty.
+func createCertificateTemplate(serverName string, uniqueSiteNameID []byte) (*x509.Certificate, error) {
+	if len(serverName) == 0 {
+		return nil, fmt.Errorf("%w: server name is empty", ErrInvalidArgument)
+	}
+
+	if len(uniqueSiteNameID) == 0 {
+		return nil, fmt.Errorf("%w: unique site name id is empty", ErrInvalidArgument)
+	}
+
+	template := &x509.Certificate{
+		SerialNumber: new(big.Int).SetBytes(uniqueSiteNameID),
+		Subject: pkix.Name{
+			CommonName:   serverName,
+			Organization: []string{"station dynamically generated"},
+		},
+		NotBefore: time.Now(),
+		NotAfter:  time.Now().AddDate(0, 0, 1),
+		DNSNames:  []string{serverName},
+	}
+
+	return template, nil
+}
+
+// generateSignedCertificate creates a certificate and then signs it using the provided Certificate Authority (CA).
```

Comment:
it is not explicit how you provide the CA to this function. The CA is read from `mkcertCARootPath`.

Reply:
it's done just after: // It uses root certificate and private key from the default location.
I don't want to mention the mkcertCARootPath because it will be cleaned in a future PR.